server/src/main/java/org/elasticsearch/action/admin/indices/forcemerge/TransportForceMergeAction.java

```diff
@@ -92,12 +93,16 @@ protected void shardOperation(
         ActionListener<TransportBroadcastByNodeAction.EmptyResult> listener
     ) {
         assert (task instanceof CancellableTask) == false; // TODO: add cancellation handling here once the task supports it
-        threadPool.executor(ThreadPool.Names.FORCE_MERGE).execute(ActionRunnable.supply(listener, () -> {
+        SubscribableListener.<IndexShard>newForked(l -> {
             IndexShard indexShard = indicesService.indexServiceSafe(shardRouting.shardId().getIndex())
                 .getShard(shardRouting.shardId().id());
-            indexShard.forceMerge(request);
-            return EmptyResult.INSTANCE;
-        }));
+            indexShard.ensureMutable(l.map(unused -> indexShard));
+        }).<EmptyResult>andThen((l, indexShard) -> {
+            threadPool.executor(ThreadPool.Names.FORCE_MERGE).execute(ActionRunnable.supply(l, () -> {
+                indexShard.forceMerge(request);
```

Comment:
nit Quickly checking, I believe a force merge does not modify the `Engine#getLastWriteNanos()`, right? I'm asking because if we force merge a hollow shard, we will unhollow it. If the last write timestamp is not updated, it will be considered as hollowable immediately by `isHollowableIndexShard()`. And I'm wondering if that's desirable. It may be OK since the shard hasn't actually received any ingestion. But feel free to comment if you think otherwise.

Reply:
No, that's only modified by actual indexing/delete ops.

Reply:
This makes me wonder if relying on the ephemeral `Engine#getLastWriteNanos()` is the right approach to evaluate if a shard hollowable, I'm thinking about degenerate cases where an inactive shard keeps moving between nodes before the TTL expires, in that case we won't hollow the shard never, right? (but this is a bit orthogonal to this discussion).

>  One argument is that we'd avoid hollowing it immediately after unhollowing it, which sounds counter-intuitive and dangerous.

Why do you think that's counter-intuitive and dangerous? In any case, if we think it'll be safer I'll introduce the tracking.

Reply:
Maybe it is not necessary. Let us discuss tomorrow before you do any additional effort here.

Reply:
After some thinking, I think I was over-thinking :)

Indeed, if there's been no ingestion, us forcing us a force merge should not necessarily leave the shard unhollow afterwards. There'll probably not be any more indexing, and the search shards can happily serve searches on optimized segments, while the indexing shard can become hollow again.

So scratch my comment. We can do it if we see it's necessary for some reason in the future. Resolving this.

Reply:
I think it is OK since ensureMutable() forks to generic pretty quickly. The comment `// just for coordination work` in the constructor could be updated to mention "ensure mutability" too.

We should also add some assertions about the threading in ensureMutable() once we have a better idea how/where we use it.